File: RadixSort.java
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Scanner;

public class main {
	private static Scanner inFile;
	private static LinkedStack myStack = new LinkedStack();
    private static int largestNum = 0;
	public static void main(String[] argv) throws IOException{ 
		try {
			if(argv.length==0) throw new FileNotFoundException("No File Found");
			File inputFile = new File(argv[0]);
			inFile = new Scanner(inputFile);
			while (inFile.hasNext() ) {
				String word=inFile.next();//this may seem like a wasted line, but without it i had an infinite loop.
				largestNum=findMax(Integer.parseInt(word),largestNum);
				myStack.push(word);
			}
			inFile.close();
			PrintWriter outFile = new PrintWriter(argv[1]);
			//initialzie RadixSort and set total digits
			RadixSort radix = new RadixSort();
			int total = findTotalDigits(largestNum);
			radix.setTotalDigits(total);
          
			//get current digit from Node, hash and add to tail of queue
			String fromStackTop = myStack.pop();
			while(true){
				String myData = fromStackTop;
				String myDigit = myData.substring(myData.length()-1);
              
				int index = radix.getCurrentHashTable().hashFunc(myDigit);              
				radix.getCurrentHashTable().getMyHash(index).addTail(fromStackTop);
				outFile.println();
				outFile.println(myData+" is sent to queue "+index);
				radix.getCurrentHashTable().getMyHash(index).printQueue(outFile);
				fromStackTop = myStack.pop();
				if(myStack.isEmpty()) break;
			}
			//print out the 10 buckets of the hashtable
			outFile.println();
			outFile.println("Current HashTable");
			outFile.println();
			radix.getCurrentHashTable().printTable(outFile);
          
			radix.increaseCurrentDigit();
			radix.switchTables();
          
			//5.0:
			while(radix.getTotalDigits() > radix.getCurrentDigit()){
				//outFile<<endl<<"Transferring data to next HashTable with new CurrentDigit"<<endl;
				radix.resetCurrentTable(); 
				for(int i =0;i < 10;i++){
					radix.getPreviousHashTable().getMyHash(i).deleteHead();
					while(!radix.getPreviousHashTable().getMyHash(i).isEmpty()){
						String tempData = radix.getPreviousHashTable().getMyHash(i).deleteHead();
						int index = radix.getCurrentHashTable().getIndex(tempData, radix.getCurrentDigit());
						radix.getCurrentHashTable().getMyHash(index).addTail(tempData);
						outFile.println(tempData +" is sent to queue " + index);
						radix.getCurrentHashTable().getMyHash(index).printQueue(outFile);
						outFile.println();
					}
				}
             
              outFile.println();
              outFile.println("Print Current HashTable");
              radix.getCurrentHashTable().printTable(outFile);
              outFile.println();
              radix.switchTables();
              radix.increaseCurrentDigit();
			} 
			outFile.println();
            outFile.println("Data successfully Sorted");
            outFile.close();
       
		} catch (Exception e){System.out.println(e);}
          
      }
	private static int findMax(int x, int y) {
		if(x>y) return x;
		else return y;
	}
	//A recursive function to find how many digits there are in a number
	private static int findTotalDigits(int x){
	    if(x == 0) return 0;
	    else return (findTotalDigits(x/10) + 1);
	}
	
  //The catch argument is strangely neccessary. without them, i got a segmentation fault
  
}

public class Node {
    private String data;
    private Node next;
    //Constructors:
    public Node(){
        next = null;
    }
    public Node(String x){
        data = x;
        next = null;
    }
    public void setData(String x){
        data=x;
    }
    public void setNext(Node n){
        next = n;
    }
    public String getData(){
        return data;
    }
    public Node getNext(){
        return next;
    }
};

public class Queue {

    private Node head;
    private Node tail;

    public Queue(){
    	head = new Node("dummy");
    	tail=head;
    }
    
    public void addTail(String x){
            tail.setNext(new Node(x));
            tail = tail.getNext();
    }
    public String deleteHead(){
        String temp = head.getData();
        head = head.getNext();
        return temp;
    }

	boolean isEmpty(){
        return (head==null);
    }
	
    void printQueue(PrintWriter output){
        if(head!=null){
            Node temp = head;
            output.print("Head ");
            while(temp.getNext()!= null){
                String x;//This will hold the data of the next node. Store -1 if it is empty, otherwise store the data
                if(temp.getNext() == null) x="NULL";
                else x = temp.getNext().getData();
                output.print("--> ("+temp.getData()+","+x+") ");
                temp=temp.getNext();
            }
            output.print("--> Tail");
            output.println();
        } else {
            output.println("Head --> (dummy,NULL) --> Tail");
        }
    }
};


public class RadixSort {


    private HashTable hashes[]= new HashTable[2];
    private int currentTable;
    private int currentDigit;
    private int previousTable;
    private int totalDigits;

    public RadixSort(){
        currentTable = 0;
        currentDigit = 0;
        previousTable = 1;
        hashes[0] = new HashTable();
        hashes[1] = new HashTable();
    }
    public void setTotalDigits(int total){
        totalDigits = total;
    }
    public int getCurrentDigit(){
        return currentDigit;
    }
    public void increaseCurrentDigit(){
        currentDigit++;
    }
    public HashTable getCurrentHashTable(){
        return hashes[(currentTable%2)];
    }
    public HashTable getPreviousHashTable(){
        return hashes[previousTable%2];
    }
    public void switchTables(){
        currentTable=(currentTable +1)%2;
        previousTable =(previousTable + 1)%2;
    }
    public int getPreviousTable(){
        return previousTable;
    }
    public int getCurrentTable(){
        return currentTable;
    }
    public int getTotalDigits(){
        return totalDigits;
    }
    public void resetCurrentTable(){
    	hashes[currentTable] = new HashTable();
    }
 
};


public class HashTable {

    private Queue[] myHash = new Queue[10];

    public HashTable(){
        //Queue[] myHash = new Queue[size];
        for(int i =0; i<10; i++){
            myHash[i] = new Queue();//initalize all 10 hashtables
        }
    }
    
    public void printTable(PrintWriter out){
        for(int i =0;i<10;i++){
            out.println("Printing Queue "+i+":");
            myHash[i].printQueue(out);
        }
    }
    public int hashFunc(String digit){
        return Integer.parseInt(digit);
    }
    
    public Queue getMyHash(int x){
    	if(x>9){
    		System.out.print("error returning hash 0");
    		return myHash[0];
    	}
    	else return myHash[x];
    }
    public int getIndex(String data, int x){
        if(data.length()<(x+1))return 0;
        else if(data.length()==(x+1)){
            String temp = data.substring(0,1);
            return Integer.parseInt(temp);
        } else{
            String temp = data.substring((data.length()-x-1),data.length()-x);
            return Integer.parseInt(temp);
        }
    }
	
};

public class LinkedStack{
	private Node Top;
	    
	public LinkedStack(){
		Top = new Node();
	}
	    
	public String pop(){
		if(isEmpty()){
			return ("empty");
	    }else{
	    	String temp = Top.getData();
	        Top = Top.getNext();
	        return temp;
	    }
	}
	public void push(String x){
		if(isEmpty()) Top = new Node(x);
		else{
			Node temp = new Node(x);
			temp.setNext(Top);
	    	Top=temp;
	    }
	}
	public boolean isEmpty(){
	        return (Top == null);
	}
	public Node getTop(){
	        return Top;
	}
	void printStack(PrintWriter output){
		Node temp = Top;
		output.println("Top ");
	    while(temp.getNext()!= null){
	    	String x;//This will hold the data of the next node. Store NULL if it is empty, otherwise store the data
	    	if(temp.getNext() == null) x="NULL";
	    	else x = temp.getNext().getData();
	    	output.println("--> ("+temp.getData()+","+x+") ");
	    	temp=temp.getNext();
	    }
	}
}

File: RadixSort.cpp
//
//  main.cpp
//  RadixSort
//
//  Created by David Futran on 10/18/15.
//  Copyright © 2015 David. All rights reserved.
//

#include <iostream>
#include <fstream>
#include <string>

using namespace std;

class Node{
public:
    string data;
    Node* next;
public:
    //Constructors:
    Node(){
        next = NULL;
    }
    Node(string x){
        data = x;
        next = NULL;
    }
    //Destructor:
    ~Node(){
        delete next;
    }
    void setData(string x){
        data=x;
    }
    void setNext(Node n){
        next = &n;
    }
    string getData(){
        return data;
    }
    Node getNext(){
        return *next;
    }
};

class LinkedStack{
private:
    Node* Top;
public:
    LinkedStack(){}//empty
    
    string pop(){
        if(isEmpty()){
            return "empty";
        }else{
            string temp = Top->getData();
            Top = Top->next;
            
            return temp;
        }
    }
    void push(string x){
        if(isEmpty()) Top = new Node(x);
        else{
            Node* temp = new Node(x);
            temp->next=(Top);
            Top=temp;
        }
    }
    bool isEmpty(){
        return (Top == NULL);
    }
    Node* getTop() const{
        return Top;
    }
    void printStack(ofstream& output){
        Node* temp = Top;
        output<<"Top ";
        while(temp->next!= NULL){
            string x;//This will hold the data of the next node. Store NULL if it is empty, otherwise store the data
            if(temp->next == NULL) x="NULL";
            else x = temp->next->data;
            output<<"--> ("<<temp->data<<","<<x<<") ";
            temp=temp->next;
        }
    }
};

class Queue{
private:
    Node* dummy = new Node("dummy");
    Node* head= dummy;
    Node* tail= dummy;
public:
    Queue(){}//empty
    
    void addTail(string x){
        if(isEmpty()){
            head=new Node("dummy");
            tail=new Node(x);
            head->next=tail;
        }else{
            tail->next = new Node(x);
            tail = tail->next;
        }
    }
    string deleteHead(){
        string temp = head->data;
        head = head->next;
        return temp;
    }
    bool isEmpty(){
        return (head==NULL);
    }
    void printQueue(ofstream& output){
        if(head!=NULL){
            Node* temp = head;
            output<<"Head ";
            while(temp->next!= NULL){
                string x;//This will hold the data of the next node. Store -1 if it is empty, otherwise store the data
                if(temp->next== NULL) x="NULL";
                else x = temp->next->data;
                output<<"--> ("<<temp->data<<","<<x<<") ";
                temp=temp->next;
            }
            output<<"--> Tail"<<endl;
        } else {
            output<< "Head --> (dummy,NULL) --> Tail"<<endl;
        }
    }
};

class HashTable{
private:
    int size = 10;
    Queue* myHash;
public:
    HashTable(){
        myHash = new Queue[size];
        for(int i =0; i<size; i++){
            myHash[i] = *new Queue();//initalize all 10 hashtables
        }
    }
    void printTable(ofstream& out){
        for(int i =0;i<10;i++){
            out<<"Printing Queue "<<i<<":"<<endl;
            myHash[i].printQueue(out);
        }
    }
    int hashFunc(string digit){
        return stoi(digit);
    }
    
    Queue* getMyHash(){
        return myHash;
    }
    int getIndex(string data, int x){
        if(data.length()<(x+1))return 0;
        else if(data.length()==(x+1)){
            string temp = data.substr(0,1);
            return stoi(temp);
        } else{
            //string myDigit = &data.at((data.length()-x-1));
            string temp = data.substr((data.length()-x-1),1);
            return stoi(temp);
        }
    }
};

class RadixSort{
private:
    HashTable* hashes[2];
    int currentTable;
    int currentDigit;
    int previousTable;
    int totalDigits;
public:
    RadixSort(){
        currentTable = 0;
        currentDigit = 0;
        previousTable = 1;
        hashes[0] = new HashTable();
        hashes[1] = new HashTable();
    }
    void setTotalDigits(int total){
        totalDigits = total;
    }
    int getCurrentDigit(){
        return currentDigit;
    }
    void increaseCurrentDigit(){
        currentDigit++;
    }
    HashTable getCurrentHashTable(){
        return *hashes[currentTable];
    }
    HashTable getPreviousHashTable(){
        return *hashes[previousTable];
    }
    void switchTables(){
        int x=currentTable;
        currentTable=previousTable;
        previousTable = x;
    }
    int getPreviousTable(){
        return previousTable;
    }
    int getCurrentTable(){
        return currentTable;
    }
    int getTotalDigits(){
        return totalDigits;
    }
 
};
//A recursive function to find how many digits there are in a number
int findTotalDigits(int x){
    if(x == 0) return 0;
    else return (findTotalDigits(x/10) + 1);
}

int main(int argc, const char * argv[]) {
    cout<<"begin"<<endl;
    try{
        if (argc < 2) {
            cout << "usage : " << argv[0] << " <input file>  <output file>\n";
            return -1;
        }
        cout << "Number of arguments : " << argc << "\n\n";
        for (int i = 0; i < argc; ++i)
            cout << "argv["<< i << "] : " << argv[i] << "\n";
        
        ifstream infile;
        string myTextFile = argv[1];
        //I had a strange error, where i couldnt open the argv[1] file directly. Instead I had to convert it to a string and open the string
        infile.open(myTextFile);
        ofstream outFile;
        //I had this next bit of code to work with a catch argument, but i would strangely get a segmentation fault if it was in the code
        /*
        if(!infile){
            throw "Fail";
        }
        */
        outFile.open(argv[2]);
        if(!infile.is_open()) cout<<"The File is not open"<<endl;
        else{
            //start with filling up the stack
            LinkedStack* myStack;
            string n;
            int largestNum =0;
            int num;
            myStack = new LinkedStack();
            while (infile >> num) {
                if(num>largestNum) largestNum=num;
                string temp = to_string(num);
                myStack->push(temp);
            }
            outFile<<endl;
            infile.close();
            
            //finally print the stack and largest number
            outFile<<"Final Stack:"<<endl;
            myStack->printStack(outFile);
            outFile<<endl<<endl<<"The largest number in input data is: "<<largestNum<<endl<<endl;
            
            //initialzie RadixSort and set total digits
            RadixSort radix;
            int total = findTotalDigits(largestNum);
            radix.setTotalDigits(total);
            
            //get current digit from Node, hash and add to tail of queue
            string fromStackTop = myStack->pop();
            while(!(fromStackTop == "empty")){
                string myData = fromStackTop;
                string myDigit = &myData.at(myData.length()-(radix.getCurrentDigit()+1));
                
                int index = radix.getCurrentHashTable().hashFunc(myDigit);
                //Node* in = new Node(fromStackTop);
                
                radix.getCurrentHashTable().getMyHash()[index].addTail(fromStackTop);
                outFile<<myData <<" is sent to queue " <<index<<endl;
                radix.getCurrentHashTable().getMyHash()[index].printQueue(outFile);
                outFile<<endl;
                fromStackTop = myStack->pop();
            }
            //print out the 10 buckets of the hashtable
            outFile<<endl<<"Current HashTable"<<endl<<endl;
            radix .getCurrentHashTable().printTable(outFile);
            
            radix.increaseCurrentDigit();
            radix.switchTables();
            
            //5.0:
            while(radix.getTotalDigits()>radix.getCurrentDigit()){
                outFile<<endl<<"Transferring data to next HashTable with new CurrentDigit"<<endl;
                radix.getCurrentHashTable() = *new HashTable();
                for(int i =0;i < 10;i++){
                    radix.getPreviousHashTable().getMyHash()[i].deleteHead();
                    while(!radix.getPreviousHashTable().getMyHash()[i].isEmpty()){
                        string tempData = radix.getPreviousHashTable().getMyHash()[i].deleteHead();
                        int index = radix.getCurrentHashTable().getIndex(tempData, radix.getCurrentDigit());
                        radix.getCurrentHashTable().getMyHash()[index].addTail(tempData);
                        outFile<<tempData <<" is sent to queue " <<index<<endl;
                        radix.getCurrentHashTable().getMyHash()[index].printQueue(outFile);
                        outFile<<endl;
                    }
                }
                outFile<<endl<<"Print Current HashTable"<<endl<<endl;
                radix .getCurrentHashTable().printTable(outFile);
            
                radix.switchTables();
                radix.increaseCurrentDigit();
            
            }
            outFile<<endl<<"Data successuly Sorted"<<endl;
            outFile.close();
        }
    //The catch argument is strangely neccessary. without them, i got a segmentation fault
    }catch(...){
        cerr<<"Error has occured somewhere.\n";
    }
    return 0;
    
}
